File: nx_bridge_py/nx_bridge.py
"""
NX-side Python SDK for communicating with the RK3588 Lite3 robot bridge.

Wire format (UDP, binary struct):
  - SensorPacket  (RK3588 → NX, port 31001): 204 bytes,  !IIi48f
  - CommandPacket (NX → RK3588, port 31002): 248 bytes,  !II60f

Usage (dual-callback mode):
    from nx_bridge_py import NxBridge, NxJointCommand

    def on_obs(state):
        print(f"rpy={state.rpy}")

    def get_act():
        cmd = NxJointCommand()
        cmd.set_standing_pose()
        return cmd

    bridge = NxBridge(recv_port=31001, robot_ip="192.168.1.100", send_port=31002)
    bridge.start(on_obs, get_act, period_ms=20)
    # ... Ctrl+C to stop ...

Usage (single-callback mode):
    def policy(state):
        cmd = NxJointCommand()
        cmd.set_standing_pose()
        return cmd

    bridge.start(policy, period_ms=20)
"""

import socket
import struct
import threading
import time
import dataclasses
from typing import Optional, Callable, Union

# ============================================================================
# Packet format strings (network byte order '!' = big-endian)
# ============================================================================

# SensorPacket: 2×uint32 + 1×int32 + 48×float32 = 204 bytes
_SENSOR_FMT = "!IIi48f"
_SENSOR_SIZE = struct.calcsize(_SENSOR_FMT)  # 204

# CommandPacket: 2×uint32 + 60×float32 = 248 bytes
_COMMAND_FMT = "!II60f"
_COMMAND_SIZE = struct.calcsize(_COMMAND_FMT)  # 248

# Sentinel values (mirrors nx_network_codes.hpp)
SHUTDOWN_SEQUENCE = 0xFFFFFFFF
HEARTBEAT_SHUTDOWN = 0

# Default ports
DEFAULT_SENSOR_PORT = 31001   # RK3588 → NX
DEFAULT_COMMAND_PORT = 31002  # NX → RK3588


# ============================================================================
# Data classes
# ============================================================================

@dataclasses.dataclass
class NxRobotState:
    """Sensor + gamepad data received from the robot (via RK3588)."""
    seq: int = 0
    timestamp_ms: int = 0
    current_state: int = 0
    rpy: tuple = (0.0, 0.0, 0.0)           # roll, pitch, yaw [rad]
    acc: tuple = (0.0, 0.0, 0.0)           # base acceleration [m/s²]
    omega: tuple = (0.0, 0.0, 0.0)         # angular velocity [rad/s]
    joint_pos: tuple = (0.0,) * 12         # joint positions [rad]
    joint_vel: tuple = (0.0,) * 12         # joint velocities [rad/s]
    joint_tau: tuple = (0.0,) * 12         # joint torques [Nm]
    cmd_vel: tuple = (0.0, 0.0, 0.0)       # user command velocity (normalised)

    @property
    def projected_gravity(self) -> tuple:
        """Gravity vector projected into the body frame (ZYX Euler convention).

        Uses the same computation as the RK3588-side RpyToRm + RmToProjectedGravity:
          projected_gravity = R_body_to_world^T * (0, 0, -1)

        Result:
          x =  sin(pitch)
          y = -cos(pitch) * sin(roll)
          z = -cos(pitch) * cos(roll)

        This is a standard feature in legged-robot RL policies.
        """
        import math
        cp = math.cos(self.rpy[1])   # cos(pitch)
        return (
            math.sin(self.rpy[1]),          #  sin(pitch)
            -cp * math.sin(self.rpy[0]),   # -cos(pitch) * sin(roll)
            -cp * math.cos(self.rpy[0]),   # -cos(pitch) * cos(roll)
        )

    def __repr__(self):
        return (f"NxRobotState(seq={self.seq}, state={self.current_state}, "
                f"rpy=({self.rpy[0]:.2f},{self.rpy[1]:.2f},{self.rpy[2]:.2f}))")


@dataclasses.dataclass
class NxJointCommand:
    """Joint command to send back to the robot (via RK3588)."""
    joint_pos_des: list = dataclasses.field(default_factory=lambda: [0.0] * 12)
    joint_vel_des: list = dataclasses.field(default_factory=lambda: [0.0] * 12)
    kp: list = dataclasses.field(default_factory=lambda: [30.0] * 12)
    kd: list = dataclasses.field(default_factory=lambda: [1.0] * 12)
    tau_ff: list = dataclasses.field(default_factory=lambda: [0.0] * 12)

    def set_standing_pose(self, hip_x=0.0, hip_y=-0.65, knee=1.30):
        """Fill with a fixed standing pose for Lite3 (12 joints, 4 legs × 3 DOF).

        Joint order: FL_HipX, FL_HipY, FL_Knee, FR_HipX, FR_HipY, FR_Knee,
                      HL_HipX, HL_HipY, HL_Knee, HR_HipX, HR_HipY, HR_Knee
        """
        # FL  FR  HL  HR
        poses = [
            hip_x,  hip_y,  knee,   # FL
           -hip_x,  hip_y,  knee,   # FR (mirror hip_x)
            hip_x,  hip_y,  knee,   # HL
           -hip_x,  hip_y,  knee,   # HR (mirror hip_x)
        ]
        self.joint_pos_des = list(poses)
        self.joint_vel_des = [0.0] * 12
        self.kp = [30.0] * 12
        self.kd = [1.0] * 12
        self.tau_ff = [0.0] * 12

    def set_gains(self, kp=30.0, kd=1.0):
        """Set uniform PD gains for all joints."""
        self.kp = [kp] * 12
        self.kd = [kd] * 12


# ============================================================================
# NxBridge
# ============================================================================

# Callback type aliases
ObservationCallback = Callable[[NxRobotState], None]
ActionCallback = Callable[[], NxJointCommand]
PolicyCallback = Callable[[NxRobotState], NxJointCommand]


class NxBridge:
    """UDP bridge to RK3588 for Lite3 robot control.

    Manages two background threads:
      - recv: receives SensorPackets, calls observation callback
      - send: periodically calls action callback, sends CommandPackets
    """

    def __init__(self,
                 recv_port: int = DEFAULT_SENSOR_PORT,
                 robot_ip: str = "192.168.1.100",
                 send_port: int = DEFAULT_COMMAND_PORT):
        self._recv_port = recv_port
        self._robot_ip = robot_ip
        self._send_port = send_port
        self._send_addr = (robot_ip, send_port)

        self._sock_recv: Optional[socket.socket] = None
        self._sock_send: Optional[socket.socket] = None

        self._running = threading.Event()
        self._recv_thread: Optional[threading.Thread] = None
        self._send_thread: Optional[threading.Thread] = None

        # Callbacks
        self._on_obs: Optional[ObservationCallback] = None
        self._get_act: Optional[ActionCallback] = None

        # For single-callback mode
        self._single_cb_mode = False
        self._latest_obs: Optional[NxRobotState] = None
        self._obs_lock = threading.Lock()

        # Sequence
        self._seq = 0
        self._seq_lock = threading.Lock()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def start(self,
              obs_or_policy: Union[ObservationCallback, PolicyCallback],
              act: Optional[ActionCallback] = None,
              period_ms: float = 20.0):
        """Start the bridge.

        Two calling conventions:

        1. Dual-callback (obs + act separate):
           bridge.start(on_observation, get_action, period_ms=20)

        2. Single-callback (obs → act):
           bridge.start(policy_function, period_ms=20)

        Args:
            obs_or_policy: ObservationCallback (mode 1) or PolicyCallback (mode 2)
            act: ActionCallback (mode 1 only)
            period_ms: Send interval in milliseconds (default 20 → 50 Hz)
        """
        if self._running.is_set():
            return

        self._period_ms = period_ms

        if act is not None:
            # Mode 1: dual-callback
            self._single_cb_mode = False
            self._on_obs = obs_or_policy
            self._get_act = act
        else:
            # Mode 2: single-callback — obs arrives → store, send thread calls policy
            self._single_cb_mode = True
            self._on_obs = self._store_obs
            policy: PolicyCallback = obs_or_policy
            self._get_act = lambda: self._call_policy(policy)

        self._setup_sockets()
        self._running.set()
        self._seq = 0

        self._recv_thread = threading.Thread(target=self._recv_loop, daemon=True)
        self._send_thread = threading.Thread(target=self._send_loop, daemon=True)
        self._recv_thread.start()
        self._send_thread.start()

        print(f"[NxBridge] Started — listening on :{self._recv_port}, "
              f"sending to {self._robot_ip}:{self._send_port} "
              f"at {1000.0 / period_ms:.0f} Hz")

    def stop(self):
        """Stop all threads and close sockets."""
        self._running.clear()
        if self._recv_thread and self._recv_thread.is_alive():
            self._recv_thread.join(timeout=1.0)
        if self._send_thread and self._send_thread.is_alive():
            self._send_thread.join(timeout=1.0)
        self._close_sockets()
        print("[NxBridge] Stopped")

    def send_shutdown(self):
        """Send an explicit shutdown signal to RK3588, triggering JointDamping."""
        if self._sock_send is None:
            return
        pkt = struct.pack(_COMMAND_FMT,
                          SHUTDOWN_SEQUENCE, HEARTBEAT_SHUTDOWN,
                          *([0.0] * 60))
        self._sock_send.sendto(pkt, self._send_addr)
        print("[NxBridge] Shutdown sent to robot")

    # ------------------------------------------------------------------
    # Internal — socket setup
    # ------------------------------------------------------------------

    def _setup_sockets(self):
        # Recv socket — bind to local port to receive SensorPackets
        self._sock_recv = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._sock_recv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._sock_recv.settimeout(0.1)  # 100 ms timeout for clean shutdown
        self._sock_recv.bind(("", self._recv_port))

        # Send socket — sends CommandPackets to RK3588
        self._sock_send = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    def _close_sockets(self):
        for sock in (self._sock_recv, self._sock_send):
            if sock is not None:
                try:
                    sock.close()
                except Exception:
                    pass
        self._sock_recv = None
        self._sock_send = None

    # ------------------------------------------------------------------
    # Internal — single-callback helpers
    # ------------------------------------------------------------------

    def _store_obs(self, state: NxRobotState):
        with self._obs_lock:
            self._latest_obs = state

    def _call_policy(self, policy: PolicyCallback) -> NxJointCommand:
        with self._obs_lock:
            obs = self._latest_obs
        if obs is not None:
            return policy(obs)
        return NxJointCommand()  # no observation yet — zeros (RK3588 holds safe)

    # ------------------------------------------------------------------
    # Internal — recv thread
    # ------------------------------------------------------------------

    def _recv_loop(self):
        """Receive SensorPackets from RK3588, call on_obs."""
        print(f"[NxBridge] Recv thread listening on port {self._recv_port}")
        while self._running.is_set():
            try:
                data, addr = self._sock_recv.recvfrom(_SENSOR_SIZE)
            except socket.timeout:
                continue
            except OSError:
                break

            if len(data) != _SENSOR_SIZE:
                print(f"[NxBridge] Received {len(data)} bytes, "
                      f"expected {_SENSOR_SIZE} — ignoring")
                continue

            state = self._unpack_sensor(data)
            if self._on_obs:
                self._on_obs(state)

        print("[NxBridge] Recv thread stopped")

    # ------------------------------------------------------------------
    # Internal — send thread
    # ------------------------------------------------------------------

    def _send_loop(self):
        """Periodically call get_act, serialise, send to RK3588."""
        period_s = self._period_ms / 1000.0
        hz = 1000.0 / self._period_ms
        print(f"[NxBridge] Send thread running at {hz:.0f} Hz")

        while self._running.is_set():
            tick_start = time.monotonic()

            if self._get_act:
                cmd = self._get_act()
            else:
                cmd = NxJointCommand()

            # Serialise with incrementing sequence number
            floats = (list(cmd.joint_pos_des) +
                      list(cmd.joint_vel_des) +
                      list(cmd.kp) +
                      list(cmd.kd) +
                      list(cmd.tau_ff))
            with self._seq_lock:
                seq = self._seq
                self._seq += 1
            pkt = struct.pack(_COMMAND_FMT, seq, seq, *floats)

            try:
                self._sock_send.sendto(pkt, self._send_addr)
            except OSError as e:
                print(f"[NxBridge] Send error: {e}")

            # Sleep for the remainder of the period
            elapsed = time.monotonic() - tick_start
            remaining = period_s - elapsed
            if remaining > 0:
                time.sleep(remaining)

        print("[NxBridge] Send thread stopped")

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    @staticmethod
    def _unpack_sensor(data: bytes) -> NxRobotState:
        """Deserialise a SensorPacket from binary."""
        fields = struct.unpack(_SENSOR_FMT, data)
        seq, ts_ms, cur_state = fields[0], fields[1], fields[2]
        floats = fields[3:]

        return NxRobotState(
            seq=seq,
            timestamp_ms=ts_ms,
            current_state=cur_state,
            rpy=floats[0:3],
            acc=floats[3:6],
            omega=floats[6:9],
            joint_pos=floats[9:21],
            joint_vel=floats[21:33],
            joint_tau=floats[33:45],
            cmd_vel=floats[45:48],
        )

File: nx_bridge/nx_bridge.cpp
/**
 * @file nx_bridge.cpp
 * @brief Implementation of the NX-side communication bridge
 */

#include "nx_bridge_api.hpp"
#include "nx_network_codes.hpp"

#include <iostream>
#include <cstring>
#include <thread>
#include <mutex>
#include <atomic>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/timerfd.h>
#include <sys/epoll.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

// ============================================================================
// File-scope configuration — safe default PD gains
// ============================================================================
// These are the fallback gains used when the action callback returns zeros.
// They are set via NxBridge::SetDefaultGains().

static float g_default_kp = 30.0f;
static float g_default_kd = 1.0f;

// ============================================================================
// PIMPL — hides all socket and threading details from the public header
// ============================================================================

struct NxBridge::Impl {
    std::string robot_ip;
    int send_port, recv_port;

    int sock_recv = -1;
    int sock_send = -1;
    struct sockaddr_in robot_addr;

    // Thread state
    std::thread recv_thread, send_thread;
    std::atomic<bool> running{false};

    // Callbacks
    ObservationCallback on_obs;
    ActionCallback      get_act;

    // For the single-callback overload: stash the latest observation
    std::mutex latest_obs_mutex;
    NxRobotState latest_obs;
    bool has_obs = false;
    bool single_cb_mode = false;

    // Send rate
    float send_period_ms = 20.0f;

    // Sequence
    uint32_t seq = 0;

    // ========================================================================
    // Recv thread — blocks on SensorPacket, deserialises, calls callback
    // ========================================================================

    void RecvThread() {
        SensorPacket pkt;
        struct sockaddr_in from_addr;
        socklen_t from_len = sizeof(from_addr);

        std::cout << "[NxBridge] Recv thread listening on port " << recv_port << std::endl;

        while (running.load()) {
            std::memset(&pkt, 0, sizeof(pkt));
            ssize_t n = recvfrom(sock_recv, &pkt, sizeof(pkt), 0,
                                 (struct sockaddr*)&from_addr, &from_len);

            if (n < 0) continue;  // timeout or error — loop

            if (static_cast<size_t>(n) != sizeof(SensorPacket)) {
                std::cerr << "[NxBridge] Received " << n
                          << " bytes, expected " << sizeof(SensorPacket) << std::endl;
                continue;
            }

            // Network-byte-order → host for integer fields
            pkt.seq           = ntohl(pkt.seq);
            pkt.timestamp_ms  = ntohl(pkt.timestamp_ms);
            pkt.current_state = static_cast<int32_t>(ntohl(
                static_cast<uint32_t>(pkt.current_state)));

            // Copy to the public-facing struct
            NxRobotState state;
            state.seq           = pkt.seq;
            state.timestamp_ms  = pkt.timestamp_ms;
            state.current_state = pkt.current_state;
            std::memcpy(state.rpy,       pkt.rpy,       sizeof(state.rpy));
            std::memcpy(state.acc,       pkt.acc,       sizeof(state.acc));
            std::memcpy(state.omega,     pkt.omega,     sizeof(state.omega));
            std::memcpy(state.joint_pos, pkt.joint_pos, sizeof(state.joint_pos));
            std::memcpy(state.joint_vel, pkt.joint_vel, sizeof(state.joint_vel));
            std::memcpy(state.joint_tau, pkt.joint_tau, sizeof(state.joint_tau));
            std::memcpy(state.cmd_vel,   pkt.cmd_vel,   sizeof(state.cmd_vel));

            if (single_cb_mode) {
                // Store the latest observation; the send thread's get_act
                // will pick it up on the next timer tick.
                std::lock_guard<std::mutex> lock(latest_obs_mutex);
                latest_obs = state;
                has_obs = true;
            } else if (on_obs) {
                on_obs(state);
            }
        }
        std::cout << "[NxBridge] Recv thread stopped" << std::endl;
    }

    // ========================================================================
    // Send thread — periodically calls get_act, serialises, sends to RK3588
    // ========================================================================

    void SendThread() {
        long period_ns = static_cast<long>(send_period_ms * 1000000.0);

        // timerfd + epoll at the configured rate (default 50 Hz)
        int tfd = timerfd_create(CLOCK_MONOTONIC, 0);
        if (tfd < 0) {
            std::cerr << "[NxBridge] timerfd_create failed" << std::endl;
            return;
        }

        struct itimerspec time_intv;
        time_intv.it_value.tv_sec  = 0;
        time_intv.it_value.tv_nsec = period_ns;
        time_intv.it_interval.tv_sec  = 0;
        time_intv.it_interval.tv_nsec = period_ns;
        timerfd_settime(tfd, 0, &time_intv, nullptr);

        int efd = epoll_create1(0);
        if (efd < 0) { close(tfd); return; }

        struct epoll_event ev;
        ev.data.fd = tfd;
        ev.events  = EPOLLIN;
        epoll_ctl(efd, EPOLL_CTL_ADD, tfd, &ev);

        struct epoll_event events[1];
        std::cout << "[NxBridge] Send thread running at "
                  << (1000.0 / send_period_ms) << " Hz" << std::endl;

        while (running.load()) {
            int n = epoll_wait(efd, events, 1, -1);
            if (n < 0) continue;

            uint64_t val;
            read(tfd, &val, sizeof(val));  // consume timer event

            // Obtain the next command from the developer's callback
            NxJointCommand cmd;
            std::memset(&cmd, 0, sizeof(cmd));
            if (get_act) {
                cmd = get_act();
            }

            // Apply safe defaults for any zero gains
            for (int i = 0; i < 12; ++i) {
                if (cmd.kp[i] <= 0.0f) cmd.kp[i] = g_default_kp;
                if (cmd.kd[i] <= 0.0f) cmd.kd[i] = g_default_kd;
            }

            // Serialise and send
            CommandPacket pkt;
            std::memset(&pkt, 0, sizeof(pkt));
            pkt.seq       = htonl(seq++);
            pkt.heartbeat = htonl(seq);
            std::memcpy(pkt.joint_pos_des, cmd.joint_pos_des, sizeof(pkt.joint_pos_des));
            std::memcpy(pkt.joint_vel_des, cmd.joint_vel_des, sizeof(pkt.joint_vel_des));
            std::memcpy(pkt.kp,             cmd.kp,             sizeof(pkt.kp));
            std::memcpy(pkt.kd,             cmd.kd,             sizeof(pkt.kd));
            std::memcpy(pkt.tau_ff,         cmd.tau_ff,         sizeof(pkt.tau_ff));

            sendto(sock_send, &pkt, sizeof(pkt), 0,
                   (struct sockaddr*)&robot_addr, sizeof(robot_addr));
        }

        close(efd);
        close(tfd);
        std::cout << "[NxBridge] Send thread stopped" << std::endl;
    }
};

// ============================================================================
// Public API
// ============================================================================

NxBridge::NxBridge(int recv_port, const std::string& robot_ip, int send_port)
    : impl_(std::make_unique<Impl>())
{
    impl_->robot_ip  = robot_ip;
    impl_->send_port = send_port;
    impl_->recv_port = recv_port;
}

NxBridge::~NxBridge() {
    Stop();
}

void NxBridge::Start(ObservationCallback on_obs, ActionCallback get_act, float period_ms) {
    if (impl_->running.load()) return;

    impl_->on_obs = std::move(on_obs);
    impl_->get_act = std::move(get_act);
    impl_->send_period_ms = period_ms;
    impl_->single_cb_mode = false;

    // --- Create recv socket ---
    impl_->sock_recv = socket(AF_INET, SOCK_DGRAM, 0);
    if (impl_->sock_recv < 0) {
        std::cerr << "[NxBridge] Failed to create recv socket" << std::endl;
        return;
    }
    int reuse = 1;
    setsockopt(impl_->sock_recv, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));

    struct timeval tv;
    tv.tv_sec  = 0;
    tv.tv_usec = 100000;  // 100 ms timeout for clean thread shutdown
    setsockopt(impl_->sock_recv, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

    struct sockaddr_in recv_addr;
    std::memset(&recv_addr, 0, sizeof(recv_addr));
    recv_addr.sin_family      = AF_INET;
    recv_addr.sin_port        = htons(impl_->recv_port);
    recv_addr.sin_addr.s_addr = INADDR_ANY;
    if (bind(impl_->sock_recv, (struct sockaddr*)&recv_addr, sizeof(recv_addr)) < 0) {
        std::cerr << "[NxBridge] Failed to bind recv socket to port "
                  << impl_->recv_port << std::endl;
        close(impl_->sock_recv);
        impl_->sock_recv = -1;
        return;
    }

    // --- Create send socket ---
    impl_->sock_send = socket(AF_INET, SOCK_DGRAM, 0);
    if (impl_->sock_send < 0) {
        std::cerr << "[NxBridge] Failed to create send socket" << std::endl;
        close(impl_->sock_recv);
        impl_->sock_recv = -1;
        return;
    }
    std::memset(&impl_->robot_addr, 0, sizeof(impl_->robot_addr));
    impl_->robot_addr.sin_family = AF_INET;
    impl_->robot_addr.sin_port   = htons(impl_->send_port);
    impl_->robot_addr.sin_addr.s_addr = inet_addr(impl_->robot_ip.c_str());

    // --- Launch threads ---
    impl_->running.store(true);
    impl_->seq = 0;
    impl_->recv_thread = std::thread(&Impl::RecvThread, impl_.get());
    impl_->send_thread = std::thread(&Impl::SendThread, impl_.get());

    std::cout << "[NxBridge] Started — listening on :" << impl_->recv_port
              << ", sending to " << impl_->robot_ip << ":" << impl_->send_port
              << " at " << (1000.0f / period_ms) << " Hz" << std::endl;
}

void NxBridge::Start(std::function<NxJointCommand(const NxRobotState&)> policy, float period_ms) {
    // Single-callback mode: recv stores latest obs (thread-safe),
    // send calls policy with that obs.

    Impl* impl = impl_.get();
    impl->single_cb_mode = true;

    ObservationCallback store_obs = [impl](const NxRobotState& state) {
        std::lock_guard<std::mutex> lock(impl->latest_obs_mutex);
        impl->latest_obs = state;
        impl->has_obs = true;
    };

    ActionCallback wrapped_get_act = [impl, policy = std::move(policy)]() -> NxJointCommand {
        NxRobotState obs;
        bool valid = false;
        {
            std::lock_guard<std::mutex> lock(impl->latest_obs_mutex);
            if (impl->has_obs) {
                obs = impl->latest_obs;
                valid = true;
            }
        }
        if (valid) {
            return policy(obs);
        }
        // No observation yet — return zeros (RK3588 holds current position safely)
        NxJointCommand cmd;
        std::memset(&cmd, 0, sizeof(cmd));
        return cmd;
    };

    Start(std::move(store_obs), std::move(wrapped_get_act), period_ms);
}

void NxBridge::Stop() {
    impl_->running.store(false);
    if (impl_->recv_thread.joinable()) impl_->recv_thread.join();
    if (impl_->send_thread.joinable()) impl_->send_thread.join();
    if (impl_->sock_recv >= 0) { close(impl_->sock_recv); impl_->sock_recv = -1; }
    if (impl_->sock_send >= 0) { close(impl_->sock_send); impl_->sock_send = -1; }
}

void NxBridge::SendShutdown() {
    if (impl_->sock_send < 0) return;

    CommandPacket pkt;
    std::memset(&pkt, 0, sizeof(pkt));
    pkt.seq       = htonl(SHUTDOWN_SEQUENCE);
    pkt.heartbeat = htonl(HEARTBEAT_SHUTDOWN);

    sendto(impl_->sock_send, &pkt, sizeof(pkt), 0,
           (struct sockaddr*)&impl_->robot_addr, sizeof(impl_->robot_addr));

    std::cout << "[NxBridge] Shutdown sent to robot" << std::endl;
}

void NxBridge::SetDefaultGains(float kp, float kd) {
    g_default_kp = kp;
    g_default_kd = kd;
}
